tool/src/org/antlr/v4/tool/GrammarTransformPipeline.java

```diff
@@ -509,4 +520,332 @@ public GrammarRootAST extractImplicitLexer(Grammar combinedGrammar) {
 		return lexerAST;
 	}
 
+	/**
+	 * Create an error listener for .stg action template group files provided via grammar options or via the command line.
+	 */
+	private STErrorListener createActionTemplateErrorListener(GrammarAST ast, STGroupFile actionTemplateGroupFile) {
+		return new STErrorListener() {
+			private final ErrorManager errorManager = ast.g.tool.errMgr;
+
+			@Override
+			public void compileTimeError(STMessage stMessage) {
+				errorManager.toolError(
+					ErrorType.ERROR_COMPILING_ACTION_TEMPLATES_FILE,
+					actionTemplateGroupFile.fileName,
+					stMessage.toString());
+			}
+
+			@Override
+			public void runTimeError(STMessage stMessage) {
+				errorManager.toolError(
+					ErrorType.ERROR_RENDERING_ACTION_TEMPLATES_FILE,
+					actionTemplateGroupFile.fileName,
+					stMessage.toString());
+			}
+
+			@Override
+			public void IOError(STMessage stMessage) {
+				reportInternalError(stMessage);
+			}
+
+			@Override
+			public void internalError(STMessage stMessage) {
+				reportInternalError(stMessage);
+			}
+
+			private void reportInternalError(STMessage stMessage) {
+				errorManager.toolError(ErrorType.INTERNAL_ERROR, stMessage.cause, stMessage.toString());
+			}
+		};
+	}
+
+	/**
+	 * Create an error listener for action templates embedded inside a grammar's actions and semantic predicates.
+	 */
+	private STErrorListener createGrammarErrorListener(GrammarAST ast) {
```

Comment:
One thing to note about the code in this error listener, which mostly revolves around translating positions - it would probably be a good idea to tidy this up a bit by exposing the information that we need in the STMessage classes in https://github.com/antlr/stringtemplate4 instead, especially `getSTRuntimeMessageSourceLocation`, which is essentially a duplicate of `STRuntimeMessage#getSourceLocation` that doesn't call `.toString` before returning.

Perhaps it would make sense to add a `Coordinate getSourceCoordinate()` method to `STMessage` (which may return null).